oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/MultithreadedTraverseWithSortStrategy.java

```diff
@@ -408,6 +411,170 @@ public void run() {
+        public File call() {
+            log.info("performing merge for {} with size {}", mergedFile.getName(), mergeTarget.size());
+            try {
+                for (int mergeFailureCount = 0; mergeFailureCount <= failureThreshold; mergeFailureCount++) {
+                    if (merge(mergeTarget, mergedFile)) {
+                        log.info("merge complete for {}", mergedFile.getName());
+                        return mergedFile;
+                    }
+                }
+                log.error("merge failed for {}", mergedFile.getName());
+            } finally {
+                mergeTaskPhaser.arriveAndDeregister();
+            }
+
+            return mergedFile;
+        }
+    }
+
+    /**
+     * Class responsible for -
+     * <ol>
+     *     <li>Watching {@link #sortedFiles} for new sorted files</li>
+     *     <li>Submitting those files in batch to an {@link ExecutorService}</li>
+     *     <li>Collecting the results (sorted files) created by those tasks</li>
+     *     <li>Merge the result with any left over files to create a single sorted file</li>
+     * </ol>
+     * Strategy -
+     * <ol>
+     *      <li>Wait for n files (compare with merged list)</li>
+     *      <li>construct new list of files to be merged by checking if its already merged</li>
+     *    and create intermediate merge file
+     *    (if final merge) merge all intermediate merge files and create sorted file
+     *      <li>add all merged files to merged list</li>
+     * </ol>
+     */
+    private class MergeRunner implements Runnable {
+        private final ArrayList<File> mergedFiles = new ArrayList<File>();
+        private final File sortedFile;
+        private int nextMergedLength = 0;
+        private final ExecutorService executorService;
+        private final int threadPoolSize = Integer.getInteger(PROP_MERGE_THREAD_POOL_SIZE, DEFAULT_NUMBER_OF_MERGE_TASK_THREADS);
+        private final int batchMergeSize = Integer.getInteger(PROP_MERGE_TASK_BATCH_SIZE, DEFAULT_NUMBER_OF_FILES_PER_MERGE_TASK);
+
+        public MergeRunner(File sortedFile) {
+            this.sortedFile = sortedFile;
+            this.executorService = Executors.newFixedThreadPool(threadPoolSize);
+        }
+
+        private List<File> getSmallestUnmergedFiles(int size) {
+            ArrayList<File> unmergedFiles = new ArrayList<File>();
+            for (File f : sortedFiles) {
+                if (!mergedFiles.contains(f) && f != MERGE_POISON_PILL) {
+                    unmergedFiles.add(f);
+                }
+            }
+            unmergedFiles.sort(new SizeFileComparator());
+
+            return unmergedFiles.subList(0, (size - 1) > unmergedFiles.size() ? unmergedFiles.size() : (size - 1));
+        }
+
+        @Override
+        public void run() {
+            nextMergedLength += batchMergeSize;
+            Phaser mergeTaskPhaser = new Phaser(1);
+            List<Future<File>> results = new ArrayList<>();
+
+            while (true) {
```

Comment:
I think we can avoid this busy wait by converting `sortedFiles` to a [BlockingQueue](https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/BlockingQueue.html) . The TraverseAndSort tasks keep on adding files to that queue and this thread consumes from that blocking queue and keeps on adding files to a local list.
If the list size grows beyond the batch size, it is sorted, merged and cleared. WDYT?
We already use a blocking queue which can be taken as an example - `taskQueue`

Reply:
changed